[Woo POS] Coupons: Search - handle cancellation in coupon strategy calls

staskus opened the pull request:
Coupon synchronization and search calls continue and finish completion even if the task is cancelled.

Add withTaskCancellationHandler to set a flag when task is cancelled and prevent completion from being called.

<!-- Remember about a good descriptive title. -->

<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

POS Coupons Strategy method calls use `withCheckedThrowingContinuation` + inner completion handlers, which do not react to `Task` being cancelled from outside. Even when the search task is canceled when typing, the completion handlers continue, eventually resulting in `controller` updating `itemViewState` with the results of the previous request.

### Solution

Use `async/await` networking methods in `CouponsRemote` that automatically handle task cancellation.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Open POS
2. Select Coupons
3. Search
4. Start typing
5. Confirm that the loading only finishes once

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->

Tested on iPad Air 11inch M3

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

### Before

https://github.com/user-attachments/assets/7b5ec11f-aa73-4446-be38-b6949bb4aef2

### After

https://github.com/user-attachments/assets/26e11ff5-2327-492a-8fd9-40eb0afb308a

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

staskus committed:
Add task cancellation handling to POS coupon strategies

Coupon synchronization and search calls continue and finish completion even if the task is cancelled.

Add withTaskCancellationHandler to set a flag when task is cancelled and prevent completion from being called.

staskus committed:
Merge branch 'trunk' into woomob-406-woo-pos-coupons-handle-search-request-cancellation

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/243">22.3</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald commented on Yosemite/Yosemite/PointOfSale/Coupons/PointOfSaleCouponFetchStrategy.swift:
```diff
@@ -99,22 +113,38 @@ public struct PointOfSaleSearchCouponFetchStrategy: PointOfSaleCouponFetchStrate
+        let cancellationState = CancellationState()
+        return try await withTaskCancellationHandler(operation: {
+            return try await withCheckedThrowingContinuation { continuation in
+                couponStoreMethods.searchCoupons(
```
as above

joshheald commented on Yosemite/Yosemite/PointOfSale/Coupons/PointOfSaleCouponFetchStrategy.swift:
```diff
@@ -45,21 +45,35 @@ private extension PointOfSaleDefaultCouponFetchStrategy {
+        let cancellationState = CancellationState()
+        return try await withTaskCancellationHandler(operation: {
+            try await withCheckedThrowingContinuation { continuation in
+                couponStoreMethods.synchronizeCoupons(
```
Would it be easier to make this function `async` and call it from within a Task in the main app? It doesn't have to retrieve anything from storage, only upsert, so it might be a quick win to convert it.

staskus replied:
That's a good question. Let me try 👍

staskus replied:
Storage methods still use `upsertSearchResultsInBackground`, which suffers from the same completionHandler issue.  

I'm sure there _should_ be a better way to handle the cancellation state when wrapping old completion handler methods. However, I fail to find it so far.

joshheald reviewed (approved):
Works well (now that I'm actually testing the right branch 🤦)

Nice one on navigating the fun of task management – I really need to get my head around the Swift 6 stuff still, so I don't want to judge too much what you've done here... but:

... there's quite a bit of complexity involved in reusing the existing methods here. Particularly for search, it might be quicker, and safer, to just do a remote search as we do in products, with an async call to the Remote.

Since the Storage-involved version that you're reusing is still really only a remote search which upserts its results into storage, it never reads from storage, perhaps you could replace it with a full async chain quite easily?

staskus committed:
Merge branch 'trunk' into woomob-406-woo-pos-coupons-handle-search-request-cancellation

staskus commented:
@joshheald, yes, we can do async/await.

As I understand, the key to supporting cancellation is to use a `Task` implementation that supports cancellation fully up the chain. 

POS Items use `enqueueWithResponseHeaders`, which internally uses `AlamofireNetwork` `responseDataAndHeaders`, which uses `Alamofire` data task [that handles cancellation](https://github.com/Alamofire/Alamofire/blob/2e7a741fc2af29bd833eddd45ee8c461987ff250/Source/Features/Concurrency.swift#L123).

The mistake I first made is trusting this [async/await enqueue method](https://github.com/woocommerce/woocommerce-ios/blob/e3f486cd9aed7e2f757befaaf0cc38babaaa781d/Networking/Networking/Remote/Remote.swift#L237), which internally uses `withCheckedThrowingContinuation` and doesn't explicitly handle cancellation.

I started using `enqueueWithResponseHeaders` and added a correct version of `async/await enqueue`, and it allows cancellation to work for coupons, same as with products.

However, pull to refresh for coupons also started to cancel and fail the request for me. I spent quite some time looking at it, and I cannot figure out why that happens. Simply loading coupons when the POS opens works, retry works, and searching also works. Now I'm stuck on this for majority of time 😀

https://github.com/user-attachments/assets/5e203819-e3fb-4945-81d1-47d52a27237f

staskus committed:
Use async/await version of network in enqueue method

enqueue async/await version implemented a custom async method using withCheckedThrowingContinuation which doesn't handle cancelation.

Use enqueueWithResponseHeaders that utilizes async/await version networking methods which fully handle Swift concurrency, including Task cancellation.

staskus committed:
Update CouponsRemote to use async/await for some methods

staskus committed:
Update CouponStoreMethods to use async/await remote methods

staskus committed:
Update PointOfSaleCouponFetchStrategy to use async/await store methods

staskus committed:
Handle network cancellation in coupons service and controller

When cancellation happens, ignore and don't display an error

staskus committed:
Add test to confirm that items are unchanged when request is cancelled

staskus commented:
@iamgabrielma @joshheald 

I updated the Coupons solution to use `async/await` for `CouponsRemote`. Deeper in the stack, it uses `Alamofire` implementation, which handles `Task` cancellation. 

We can handle a cancellation within the `Controller` and ignore it. The same solution is used for `PurchasableItems`.

staskus committed:
Remove leftover code - CancellationState

staskus committed:
Merge branch 'trunk' into woomob-406-woo-pos-coupons-handle-search-request-cancellation

joshheald commented on Networking/Networking/Remote/Remote.swift:
```diff
@@ -235,28 +234,7 @@ public class Remote: NSObject {
-                }
-            }
-        }
+        try await enqueueWithResponseHeaders(request, mapper: mapper).data
```
We should probably get eyes on any part of the codebase that's using this `enqueue` function, just to make sure we handle it properly in the layers above. We could have been relying on the way errors can be swallowed by tasks like this (at least, I think they can – the Alamofire cancellation one, for instance...?)

Looks like the remotes are:
- Blaze
- GenerativeContent
- GiftCardStats
- GoogleListings
- JustInTimeMessages
- MetaData
- Orders
- Payment (in-app purchase, probably unused)
- ProductBundle
- ProductCategories
- Products
- ProductsReports
- ShippingMethods
- StoreOnboardingTasks
- WCPay
- WordpressSite
- WordpressTheme

staskus replied:
I haven't found explicit places where cancelling was important. The networking layer itself is quite well covered by unit tests, although the test seem to cover more conventional happy and unhappy paths. I did smoke testing around these areas, but haven't jumped on any out of place errors.

joshheald commented on WooCommerce/Classes/POS/Controllers/PointOfSaleCouponsController.swift:
```diff
@@ -100,6 +100,16 @@ private extension PointOfSaleCouponsController {
                 setCouponsLoadedViewState(storedCoupons, hasMoreItems: true)
             }
 
+            await refreshFirstPage()
```
Given the way we call this, it might be better to rename it `loadFirstPageRemotely`? Though, refresh isn't inaccurate, it just seems a little backwards to call it from a load function.

joshheald commented on WooCommerce/Classes/POS/Controllers/PointOfSaleCouponsController.swift:
```diff
@@ -100,6 +100,16 @@ private extension PointOfSaleCouponsController {
+            if let couponError = error as? PointOfSaleCouponServiceError {
+                setCouponsErrorViewState(couponError)
+            }
+        }
```
Are we sure we want to silently ignore any other errors?
```suggestion
        } catch let error as PointOfSaleCouponServiceError {
            setCouponsErrorViewState(error)
        } catch {
            /// Other error handling/logging
        }
```

joshheald commented on Yosemite/Yosemite/PointOfSale/Coupons/PointOfSaleCouponFetchStrategy.swift:
```diff
@@ -45,35 +45,9 @@ private extension PointOfSaleDefaultCouponFetchStrategy {
-                await cancellationState.cancel()
-            }
-        })
+        return try await couponStoreMethods.synchronizeCoupons(siteID: siteID,
+                                                               pageNumber: pageNumber,
+                                                               pageSize: Constants.defaultPageSize)
```
💜 Nice

joshheald commented on Yosemite/Yosemite/PointOfSale/Coupons/PointOfSaleCouponService.swift:
```diff
@@ -62,9 +63,11 @@ public final class PointOfSaleCouponService: PointOfSaleCouponServiceProtocol {
         } catch {
             if try await checkRemoteStoreCouponSettings() {
-                throw error
+                throw PointOfSaleCouponServiceError.couponsLoadingError
```
losing the error detail here may make it more difficult to support customers in future... consider passing the whole error up and then unpacking it more in the view layer?

joshheald commented on Yosemite/Yosemite/Stores/CouponStore.swift:
```diff
@@ -50,10 +50,16 @@ public final class CouponStore: Store {
-                                       pageNumber: pageNumber,
-                                       pageSize: pageSize,
-                                       onCompletion: onCompletion)
+            Task {
+                do {
+                    let hasMorePages = try await methods.synchronizeCoupons(siteID: siteID,
+                                                                            pageNumber: pageNumber,
+                                                                            pageSize: pageSize)
+                    onCompletion(.success(hasMorePages))
+                } catch {
+                    onCompletion(.failure(error))
+                }
+            }
```
I'm not against this, but caution that changing the action handling in this way might make testing harder. You don't have an async chain when you dispatch this action, so you're reliant on the `onCompletion` handler being called and maybe using a `waitFor`. Those haven't played well with Tasks/checkedContinuations in the past.

Do you think it would be possible/safer to contain the async changes within `methods`? Yes, ideally the old method would call the new async one so there's still a touch of that issue, but you wouldn't neccesarily need that to work for the tests to pass.

joshheald replied:
PS I'm seeing some issues running tests locally due to threading, things falling off the main thread potentially in the `synchronizeCoupons` Task. Perhaps it just wants to be MainActor, ideally from the tests or calling its completion handlers on main

joshheald reviewed (approved):
Well done for getting it working.

I think we'll need to test the other areas that use the updated `enqueue`. There's a few other questions and suggestions inline.

staskus committed:
Rename refreshFirstPage to  loadFirstPageRemotely

staskus committed:
Handle all errors in CouponService

- Pass underlyingError together with couponsLoadingError in case we want to handle it more granularly in the future
- Handle all errors coming from the service

staskus committed:
Update tests for adding underlyingError in couponsLoadingError

staskus committed:
Move Task to CouponStoreMethods

staskus committed:
Merge branch 'trunk' into woomob-406-woo-pos-coupons-handle-search-request-cancellation

staskus committed:
Merge branch 'trunk' into woomob-406-woo-pos-coupons-handle-search-request-cancellation

staskus committed:
Use MainActor Tasks in CouponStoreMethods

staskus committed:
Update PointOfSaleDefaultCouponFetchStrategyTests.swift
